state/cluster/badger/mutator.go

```diff
@@ -218,3 +205,74 @@ func (m *MutableState) Extend(block *cluster.Block) error {
 	}
 	return nil
 }
+
+// checkDupeTransactionsInUnfinalizedAncestry checks for duplicate transactions in the un-finalized
+// ancestry of the given block, and returns a list of all duplicates if there are any.
+func (m *MutableState) checkDupeTransactionsInUnfinalizedAncestry(block *cluster.Block, includedTransactions map[flow.Identifier]struct{}, finalHeight uint64) ([]flow.Identifier, error) {
+
+	var duplicateTxIDs []flow.Identifier
+	err := fork.TraverseBackward(m.headers, block.Header.ParentID, func(ancestor *flow.Header) error {
+		payload, err := m.payloads.ByBlockID(ancestor.ID())
+		if err != nil {
+			return fmt.Errorf("could not retrieve ancestor payload: %w", err)
+		}
+
+		for _, tx := range payload.Collection.Transactions {
+			txID := tx.ID()
+			_, duplicated := includedTransactions[txID]
+			if duplicated {
+				duplicateTxIDs = append(duplicateTxIDs, txID)
+			}
+		}
+		return nil
+	}, fork.ExcludingHeight(finalHeight))
+
+	return duplicateTxIDs, err
+}
+
+// checkDupeTransactionsInFinalizedAncestry checks for duplicate transactions in the finalized
+// ancestry, and returns a list of all duplicates if there are any.
+func (m *MutableState) checkDupeTransactionsInFinalizedAncestry(includedTransactions map[flow.Identifier]struct{}, minRefHeight, maxRefHeight uint64) ([]flow.Identifier, error) {
+	var duplicatedTxIDs []flow.Identifier
+
+	// Let E be the global transaction expiry constant, measured in blocks
+	//
+	// 1. a cluster block's reference block height is equal to the lowest reference
+	//    block height of all its constituent transactions
```

Comment:
trying to increase mathematical rigour here a bit:
```suggestion
	// Let E be the global transaction expiry constant, measured in blocks. For each
	// T ∈ `includedTransactions`, we have to decide whether the transaction
	// already appeared in _any_ finalized cluster block.
	// Notation:
	//   - consider a valid cluster block C and let c be its reference block height
	//   - consider a transaction T ∈ `includedTransactions` and let t denote its
	//     reference block height
	//
	// Boundary conditions:
	// 1. C's reference block height is equal to the lowest reference block height of
	//    all its constituent transactions. Hence, C can contain T if and only if c <= t.
```